parsec/mca/termdet/local/termdet_local_module.c

```diff
@@ -0,0 +1,240 @@
+{
+    assert(&parsec_termdet_local_module.module == tp->tdm.module);
+    tp->tdm.callback = cb;
+    tp->tdm.monitor = PARSEC_TERMDET_LOCAL_NOT_READY;
+}
+
+static parsec_termdet_taskpool_state_t parsec_termdet_local_taskpool_state(parsec_taskpool_t *tp)
+{
+    if( tp->tdm.module == NULL )
+        return PARSEC_TERM_TP_NOT_MONITORED;
+    assert(tp->tdm.module == &parsec_termdet_local_module.module);
+    if( tp->tdm.monitor == PARSEC_TERMDET_LOCAL_TERMINATED )
+        return PARSEC_TERM_TP_TERMINATED;
+    if( tp->tdm.monitor == PARSEC_TERMDET_LOCAL_BUSY )
+        return PARSEC_TERM_TP_BUSY;
+    if( tp->tdm.monitor == PARSEC_TERMDET_LOCAL_NOT_READY )
+        return PARSEC_TERM_TP_NOT_READY;
+    assert(0);
+    return -1;
+}
+
+static int parsec_termdet_local_taskpool_ready(parsec_taskpool_t *tp)
+{
+    assert( tp->tdm.module != NULL );
+    assert( tp->tdm.module == &parsec_termdet_local_module.module );
+    assert( tp->tdm.monitor == PARSEC_TERMDET_LOCAL_NOT_READY );
+    parsec_atomic_cas_ptr(&tp->tdm.monitor, PARSEC_TERMDET_LOCAL_NOT_READY, PARSEC_TERMDET_LOCAL_BUSY);
+    return PARSEC_SUCCESS;
+}
+
+static int32_t parsec_termdet_local_taskpool_set_nb_tasks(parsec_taskpool_t *tp, int32_t v)
+{
+    int32_t ov, nbpa = 1; // By default we are not the one to discover nbpa gets to 0
+    PARSEC_DEBUG_VERBOSE(10, parsec_debug_output, "TERMDET-LOCAL:\tNB_TASKS -> %d", v);
+    if(tp->nb_tasks != v) {
+        do {
+            ov = tp->nb_tasks;
+        } while(! parsec_atomic_cas_int32(&tp->nb_tasks, ov, v));
+        if( ov == 0 && v > 0 ) {
+            nbpa = parsec_atomic_fetch_inc_int32(&tp->nb_pending_actions) + 1;
+        } else if( ov > 0 && v == 0 ) {
+            nbpa = parsec_atomic_fetch_dec_int32(&tp->nb_pending_actions) - 1;
+        }
+    }
+    if( tp->tdm.monitor == PARSEC_TERMDET_LOCAL_BUSY && nbpa == 0 ) {
+        PARSEC_DEBUG_VERBOSE(10, parsec_debug_output, "TERMDET-LOCAL:\tnbpa == 0, Call callback");
+        if( parsec_atomic_cas_ptr(&tp->tdm.monitor, PARSEC_TERMDET_LOCAL_BUSY, PARSEC_TERMDET_LOCAL_TERMINATED) ) {
+            tp->tdm.callback(tp);
+        }
+    }
+    return tp->nb_tasks;
+}
+
+static int32_t parsec_termdet_local_taskpool_set_runtime_actions(parsec_taskpool_t *tp, int32_t v)
+{
+    int32_t ov;
+    PARSEC_DEBUG_VERBOSE(10, parsec_debug_output, "TERMDET-LOCAL:\tNB_PA -> %d", v);
+    do {
+        ov = tp->nb_pending_actions;
+    } while(!parsec_atomic_cas_int32(&tp->nb_pending_actions, ov, v));
+    if( tp->tdm.monitor == PARSEC_TERMDET_LOCAL_BUSY && v == 0 ) {
+        PARSEC_DEBUG_VERBOSE(10, parsec_debug_output, "TERMDET-LOCAL:\tnbpa == 0, Call callback");
+        if( parsec_atomic_cas_ptr(&tp->tdm.monitor, PARSEC_TERMDET_LOCAL_BUSY, PARSEC_TERMDET_LOCAL_TERMINATED) ) {
+            tp->tdm.callback(tp);
+        }
+    }
+    return v;
+}
+
+static int32_t parsec_termdet_local_taskpool_addto_nb_tasks(parsec_taskpool_t *tp, int32_t v)
```

Comment:
In this termination detector, we use `nb_pending_actions` as the witness for termination: when it reaches 0, all work (tasks as well as runtime work) is done. When `nb_tasks > 0`, there is 1 additional `nb_pending_actions`, when `nb_tasks == 0`, there is no additional `nb_pending_actions`.

If the runtime system is done sending messages on behalf of this taskpool when the last task completes, `nb_tasks` will reach 0, and it's possible that `nb_pending_actions` then reach 0, triggering the termination for the taskpool.

Similarly, when we run the internal init for the taskpool, we set `nb_tasks` to something `>0`, and that adds 1 to `nb_pending_actions`.

I'm not sure I answered your question... It might be better to discuss about this?